### ISSUE TITLE
dataflow: DRAINING should be treated as a terminal state if lifecycle.create_before_destroy is true

### ISSUE THREAD

**Issue by n-oden** [hashicorp/terraform-provider-google#10559]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment. If the issue is assigned to the "modular-magician" user, it is either in the process of being autogenerated, or is planned to be autogenerated soon. If the issue is assigned to a user, that user is claiming responsibility for the issue. If the issue is assigned to "hashibot", a community member has claimed the issue already.

### Description

Consider a dataflow job with `lifecycle.create_before_destroy` set to `true`.  If a plan is applied that forces the replacement of the resource, terraform will first attempt to create the replacement job (presumably/hopefully there is something interpolated into the job name in order to allow the job to exist in parallel with its replacement) and then issues the drain command to the old job... and waits.  And waits.  Plan execution will not finish until the job reaches the DRAINED state, and there are multiple scenarios (mostly involving windowing) in which this could take hours, plural.  In the case of a terraform cloud or atlantis workspace, this potentially blocks any other use of the workspace for a substantial amount of time.

But for the purposes of the user, there is no reason to wait: the replacement job has been created.  The old job is in state DRAINING, and it will either reach DRAINED or FAILED depending on the local weather, but there is no way for it to revert to RUNNING.

I think it would be reasonable to treat `DRAINING` as a terminal state, delete the old job from tf state and move on at this point.

### New or Affected Resource(s)

* google_dataflow_job
* google_dataflow_flex_template_job

### Potential Terraform Configuration

As noted above, I think this should simply be the default behavior, but potentially this could also be a configurable boolean, e.g. something like:

```tf
resource "random_id" "job_suffix" {
  byte_length = 4
  keepers = {
    region = var.region
    tmpl = var.myjob_gcs_tmpl
  }
}

resource "google_dataflow_job" "myjob" {
  name = "myjob"
  template_gcs_path = var.myjob_gcs_tmpl
  region = var.region
  on_delete = "drain"
  wait_for_drain = false
  lifecycle {
    create_before_destroy = true
  }
}
```

**Comment by n-oden** [comment]
n.b. I'm happy to attempt to put together a PR for this myself, but I'd like to hear if the maintainers agree or disagree with my reasoning here.

**Comment by rileykarson** [comment]
We can't actually tell if lifecycle.create_before_destroy is set in Terraform, unfortunately, so customizing behaviour based on that isn't possible! Only a pretty limited set of information makes it to the provider.

**Comment by n-oden** [comment]
@rileykarson hm.  okay, that being the case, would you accept a PR that added an explicit `wait_for_drain` boolean parameter and some documentation about when it is and is not safe to use?

**Comment by rileykarson** [comment]
It'd be preferable if the condition were inverted i.e. `skip_drain_on_destroy` as adding new fields that default to `true` is fairly finicky but I don't see a big problem w/ that approach, no

**Comment by n-oden** [comment]
@rileykarson take a gander at https://github.com/hashicorp/terraform-provider-google-beta/pull/3868 -- if you like the shape of it I'll open another PR against the non-beta provider to match it.

**Comment by n-oden** [comment]
Reopening because we have some associated PRs to review. :)
